paper/paper.tex

```diff
@@ -0,0 +1,249 @@
+
+This paper introduce an interface called MutableArithmetics (MA).
+It allows for mutable types to implement an arithmetics exploiting their mutability and for algorithms to
+exploit their mutability while remaining completely generic.
+Moreover, it provides the following additional features:
+\begin{enumerate}
+  \item
+    \label{item:reimplement}
+    it re-implements part of the Julia standard library on top of the API to allow mutable type to use a more efficient version than the default one.
+  \item
+    \label{item:rewrite}
+    it defines a \texttt{@rewrite} macro that rewrites an expression using the standard operations (e.g \texttt{+}, \texttt{*}, ...) into a code that exploits the mutability of the intermediate values created when evalutation the expression.
+\end{enumerate}
+
+JuMP~\cite{dunning2017jump} used to have its own API for mutable operations on JuMP expressions and
+its own JuMP-specific implementation of \ref{item:reimplement} and \ref{item:rewrite}.
+These performance improvements are one of the key features of JuMP that allows it to have a performance that is competitive to commercial algebraic modeling languages that are not embedding in another language for linear programs~\cite[Section~3--4]{dunning2017jump}.
+These features were refactored into MA, generalizing them to arbitrary mutable types.
+Starting from JuMP v0.21, JuMP expression and MOI functions implement the MutableArithmetics API and
+the JuMP-specific implementations of \ref{item:reimplement} and \ref{item:rewrite} were replaced by the generic versions implemented in MA on top of the MA API.
+
+\section{Design consideration}
+This section provide concrete examples that motivates the design of MA.
+We organize the section in four subsection that describes the need of four key features of MA's API.
+
+\subsection{May mutate}
+Consider the example of the sum of the elements of a vector.
+By default, Julia \texttt{sum} function will compute with a code equivalent to the following:
+\begin{lstlisting}[language = Julia]
+function sum(x::Vector)
+    acc = zero(eltype(x))
+    for el in x
+        acc = acc + el
+    end
+    return acc
+end
+\end{lstlisting}
+If the type of the elements of \texttt{x} is \texttt{BigInt}, it is more efficient to replace the line
+\lstinline|acc = acc + el| by the line
+\lstinline|Base.GMP.MPZ.add!(acc, el)|.
+Indeed, as the operation \lstinline|+| cannot modify its arguments,
+it will need to allocate a new instance of \texttt{BigInt} to contain the result.
+On the other hand, \lstinline|Base.GMP.MPZ.add!| modifies \lstinline|acc| in place to contain the result.
+
+Even if using \lstinline|Base.GMP.MPZ.add!| provides a significant performance improvement,
+the time complexity is identical.
```

Comment:
I don't understand this time complexity argument. You mean big-O?

Perhaps the other way around: 

Do you meant something like
```Julia
The time complexity of using \lstinline|Base.GMP.MPZ.add!| is identical to \lstinline|+|,
but the former does not allocate memory, and so results in a significant performance
improvement.
```